devicemapper: remove thin pool if 'initDevmapper' failed

coolljt0725 opened the pull request:
if initDevmapper failed after creating thin-pool, the thin-pool will not be removed,
this would cause we can't use the same lvm to create another thin-pool.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

coolljt0725 commented:
ping @rhvgoyal

thaJeztah commented:
Just to be sure (didn't check the entire flow of the code), but removing will _only_ happen if it was just created, not in some other cases? (was thinking of cases where "an error" occurred, and it would remove an existing thin pool)

coolljt0725 commented:
@thaJeztah  this will only remove the pool created by the current starting, if there is a existing thin pool, it should not get to there.  Use following steps can reproduce this issue:
1. start docker with devicemapper+direct lvm
2. stop docker and restart docker with devicemapper+loop, this will failed to start and left a thin pool behind.
3. restart the docker with devicemapper+direct lvm， this will failed to start because the thin pool created by step 2 exists.

thaJeztah commented:
Thanks for clarifying @coolljt0725!

rhvgoyal commented:
@coolljt0725 what was the error you faced when you switched from direct lvm thin pool to loop device pool.

So IIUC, docker is using lvm thin pool, then you change the configuration to make it use loop devices and that fails. And now you want same docker to go back to using lvm thin pool?

I don't mind removing thin pool if error happens but above configuration change is not a good idea.

coolljt0725 commented:
@rhvgoyal  we face `UUID verifying failed` error.  

> So IIUC, docker is using lvm thin pool, then you change the configuration to make it use loop devices and that fails. And now you want same docker to go back to using lvm thin pool?

yes

coolljt0725 committed:
devicemapper: remove thin pool if 'initDevmapper' failed

if initDevmapper failed after creating thin-pool, the thin-pool will not be removed,
this would cause we can't use the same lvm to create another thin-pool.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

cpuguy83 commented:
This seems ok to me, @rhvgoyal LGTY?

rhvgoyal commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1840,6 +1840,14 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 			return err
 		}
+		defer func() {
+			if retErr != nil {
```
Does this code really work? I see that retErr is named return variable but nobody is operating on that variable. So it should always be 0 and defer function will not do anything?

Or I am understanding it wrong.

coolljt0725 replied:
This works. The `return err` will assign the value to `retErr` https://tour.golang.org/welcome/1

rhvgoyal commented:
Should we extend it to get rid of lvm thin pool setup by docker if an error happens after setting up thin pool. 

Also initDevMapper() has become too big and we need to really break it down.

rhvgoyal commented:
@coolljt0725 ok, thanks for clarifying retErr assignment. 

LGTM

BTW, will be good if we can implement cleanup logic for lvm thin pool as well. May be in a separate PR.

cpuguy83 reviewed (approved):
LGTM
